text/0000-eliminate-finalize.md

```diff
@@ -0,0 +1,142 @@
+
+### Verifying the proof
+
+The following 4 steps are necessary for the sender to prove payment to the receiver's address:
+
+1. Provide the kernel and show that it was confirmed on-chain.
+2. Prove knowledge of `sender.nonce`
+3. Show that `sender.public_nonce + receiver.public_nonce + (proof_nonce * G)` = `total_nonce * G` (ie. the `k*G` in the kernel signature)
+4. Provide the preimage to `proof_nonce` (`diffie_secret | receiver.public_excess | tx.amount`)
+
+## Slate Format
+
+A slate is now only needed for passing information from the sender to receiver. It must include the following data:
+
+* `tx.amount`
+* `tx.offset`
+* `tx.kernel_features`
+* `tx.fee`
+* `tx.inputs`
+* `receiver.change_output`
+* `receiver.public_excess` & `receiver.public_nonce` (to prevent the need for grinding)
+* `sender.public_excess`, `sender.public_nonce`, & `sender.partial_sig`
+* `time_to_live` (optional)
+
+TODO: Define JSON & binary formats
+TODO: Mention QR codes
+
+## Backward Compatibility
+
+TODO: Decide how to handle versioning. In theory, this should be simple. If given an old slatepack address, use the old slate format. If given a new, one-time address, then the new slate format is supported.
+
+## Invoices
+
+To send an invoice to a payer, the recipient provides the payer with a newly generated one-time address, an amount, and any additional fields we decide to support (memo, refund address, expiration, etc). The payer, upon approving payment of the invoice, simply follows the standard 2-step send process described in this RFC. There's no need for a special workflow.
+
+## Security
+
+### Nonce Reuse
+
+Nonces should be randomly generated by the receiving wallet, rather than derived from seed, since reuse of a nonce could lead to loss of funds. Receiving wallets should keep track of all one-time addresses (excesses & nonces) they generate in a list (`address_list`), along with an indicator for whether they have been used yet.
+
+Upon receipt of a partial transaction (slate), the receiver should obtain a lock on `address_list` and then check whether the address has been used before. If it has not, it's safe to finalize the transaction, but before broadcasting, the address should be marked as used in the `address_list`, and then the lock can be safely released.
+
+### Play Attacks
+
+TODO: Describe play attacks, and finish this section.
+
+Any inputs sent from a wallet to a one-time address should be considered as sent. They should not be cancelable, however, the sender should have the option to add additional inputs and/or increase the fee (equivalent of bitcoin's RBF).
+
+Receiver should be unable to claim a problem with the transaction. If they do, it's trivial to show them the slate that was sent, and prove that it's a valid partial transaction containing everything necessary for the receiver to finalize.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Addresses are significantly longer (170 characters).
```

Comment:
I didn't quite get it, will there still be a single address? I.e. does this change the `grin1` slatepack address? Or will the slatepack address and the one-time use address of the receiver co-exist as two separate addresses?

Reply:
Single address. This changes the existing `grin1` slatepack address. The example above shows the values that were combined to create the address (version, ed25519 pubkey, pub excess, pub nonce). That's more of just a test vector. The actual address that you share would be the last one: `grin1qx87jxj3htwwhj5k6hy0xhd0fszpntznz26clfjfdsmpggy437gt5qmrjkfvc02qdgth86ke3dv65rz5cy8xgyhdln0h33zl4akgzrvs6spn96dt6en2hyvmyfsnulmrxt6ynjddk34nywcdq6v4874t79w063gugaa22`

Reply:
Ahh, okay got it!